src/estimators/independent_set_estimator.py

```diff
@@ -0,0 +1,112 @@
+
+    |A union B| = |A| + |B| - |A| * |B| / |U|.
+
+  Note that the estimators for A and B will in general return cumulative histograms,
+  which may contain values for frequencies greater than 1.  The
+  IndependentSetEstimator also returns a cumulative histogram, which estimates the
+  number of items in the intersection at various frequencies.
+  """
+
+  def __init__(self, single_sketch_estimator, universe_size):
+    """Instantiates an IndependentSetEstimator object.
+
+    Args:
+      single_sketch_estimator:  An object derived from EstimatorBase for
+        providing cardinality estimates of individual sketches.  The
+        single_sketch_estimator is assumed to return a cumulative histogram.
+      universe_size: Size of the universe.
+    """
+    self.single_sketch_estimator = single_sketch_estimator
+    self.universe_size = universe_size
+
+  def __call__(self, sketch_list):
+    """Computes the frequency histogram of a union of sketches.
+
+    Args:
+      sketch_list: A list of sketches.  For each sketch in the list, its
+        cumulative frequency histogram is determined using the 
+        single_sketch_estimator that was provided in the 
+        IndependentSetEstimator constructor.
+
+    Returns:
+      A cumulative histogram h.  The element h[i] represents the number
+      of items in the union having frequency i+1 or greater, under the
+      independence assumption.
+    """
+    if not sketch_list:
+      return [0]
+    
+    a_hist = [0]  # Exact frequencies of current estimate
+    for sketch in sketch_list:
+      # ch is a cumulative histogram.
+      ch = self.single_sketch_estimator([sketch])
+      # b_hist is the exact frequency histogram for the new sketch.
+      b_hist = [ch[i] - ch[i+1] for i in range(len(ch)-1)] + [ch[-1]]
+
+      # To compute the histogram of the union, we look at all pairs
+      # A[i] and B[j], where A[i] (resp. B[j]) is the number of items with
+      # frequency i+1 in A (resp. j+1 in B).  Items in the intersection
+      # of A[i] and B[j] will have frequency i+j+2 in the union.
+      # These items will need to be deducted from the frequency counts for
+      # both i and j to avoid overcounting.
+      c_hist = a_hist.copy() + [0] * (len(b_hist) + 1)
+      for i in range(len(b_hist)):
+        c_hist[i] += b_hist[i]
+      for i in range(len(a_hist)):
+        for j in range(len(b_hist)):
+          overlap = a_hist[i] * b_hist[j] / self.universe_size
+          if overlap:
+            c_hist[i] -= overlap
+            c_hist[j] -= overlap
+            c_hist[i+j+1] += overlap
+
+      a_hist = c_hist
+
+      assert sum(a_hist) <= self.universe_size, (
+        "Constraint violation: sketch is larger than universe")
+
+    # At this point, a_hist is the histogram of exact frequencies,
+    # e.g., a_hist[i] is the number of items having frequency exactly i+1.
+    # Now convert it into a histogram of cumulative frequencies.
+
+    a_hist = [int(f) for f in a_hist]
```

Comment:
Do we have to round it here or maybe let the caller round it similar to the estimators previously written?

Reply:
I have removed the rounding.